types/util/tools.d.ts

```diff
@@ -524,22 +544,105 @@ export type WidenLiterals<T> =
+
 /**
- * Recursively Update a deep property from a given path
+ * LCDT (Lowest Common Denominator Type)
  *
- * @param Keys array of keeps into the object
- * @param U The object to change
- * @param T the next type at the key path
+ * Right preferential
+ *
+ * When the types are different, the right is always taken, eg
+ * ```typescript
+ * LCDT<string, number>; // number
+ * LCDT<number, string>; // string
+ * ```
+ *
+ * The important part happens when one types extends the other, but not the other way around.
+ * This is to say that all squares are rectangles but not all rectangles are squares
+ * ```typescript
+ * LCDT<'a', string>; // string
+ * LCDT<string, 'a'>; // string
+ * ```
+ * Also works with unions
+ * ```typescript
+ * LCDT<'a', 'a' | 'b'>; // 'a' | 'b'
+ * LCDT<'a' | 'b', 'a'>; // 'a' | 'b'
+ * LCDT<'a', 'b'>; // 'a' | 'b'
+ * LCDT<'b', 'a'>; // 'b' | 'a'
+ * ```
+ *
+ * Note:
+ * You'd think that `type LCDT<L, R> = L extends R ? (R extends L ? L : R) : R;` works, but it doesn't
+ * You have to re-infer R via `infer N` using this weird function syntax
  *
  * <created by @harris-miller>
  */
-export type DeepModify<Keys extends readonly PropertyKey[], U, T> =
-  Keys extends [infer K, ...infer Rest]
-    ? K extends keyof U
-      ? Rest extends readonly []
-        ? Omit<U, K> & Record<K, T>
-        : Rest extends readonly PropertyKey[]
-          ? Omit<U, K> & Record<K, DeepModify<Rest, U[K], T>>
-          : never
-      : never
-    : never;
+type LCDT<L, R> = (<L extends R>(v: L) => L) extends (<L extends R>(v: L) => infer N) ? N extends L ? L
+  : (L extends string ? N extends string ? L | N : N : (L extends number ? N extends number ? L | N : N : (L extends boolean ? N extends boolean ? L | N : N : N)))
+  : R;
+
+export type _MergeObjects<L extends object, R extends object> = {
+  [K in (keyof (L & R))]: {
+    0: TakeRightToLeft<L, R, K>;
+    // @ts-ignore
+    1: LCDT<L[K], R[K]>
+  }[KeyOfBoth<L, R, K>];
+};
+
+
+/**
+ * MergeObjects
+ *
+ * A better collapse of types for Merging Two Objects
+ * MergeRight is just Object.assign in code, and MergeLeft is just MergeRight with the operands flipped
+ * What this Type does is better collapse the types together
+ * eg
+ * ```typescript
+ * type Foo = { foo: string };
+ * type Foobar = { foo: string; bar: string };
+ *
+ * Object.assign({}, {} as Foobar, {} as Foo); // Foobar & bar, a redundant intersection
+ * mergeRight({} as Foobar, {} as Foo); // Foobar, since Foobar extends Foobar
+ * mergeRight({} as Foobar, {} as Foo); // also Foobar, because regardless of the direction you merge, one type extends the other
+ * ```
+ *
+ * <created by @harris-miller>
+ */
+export type MergeObjects<L extends object, R extends object> = {
+  0: _MergeDeepObjects<L, R>,
+  1: L
+}[A.Extends<L, R>];
+
+// first we text if all the values in both `L` and `R`, if true, defer to `MergeObjects` (see last line), otherwise...
```

Comment:
```suggestion
// first we test if all the values in both `L` and `R` are primitive, if true, defer to `MergeObjects` (see last line), otherwise...
```